added trimat_forward cuda kernel.

kohankhaki opened the pull request:
# [Compute] Trimat CUDA Kernel


## Type of Change
- [x] New Pocket Reference
- [ ] Edit to Existing Pocket Reference
- [ ] Other (please describe):

Fixes #

## Book

<!-- Indicate which book this PR affects -->

- [ ] fundamentals
- [ ] nlp
- [ ] cv
- [ ] rl
- [ ] fl
- [ ] responsible_ai
- [x] compute

## Description

Adding trimat CUDA kernel pocket reference.

## Checklist

- [x] I have included appropriate contributor tags (`{{#author}}` or `{{#authors}}`)
- [x] I have added the reading time preprocessor tag under the title
- [x] Content is concise and within 7 minutes reading time
- [x] I have included relevant references and further reading links
- [x] I have tested locally using `mdbook watch books/<book-name> --open`
- [x] Pre-commit hooks pass without errors
- [ ] I have linked to related issues

kohankhaki committed:
added trimat_forward cuda kernel.

nerdai commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
 
+## Introduction
+
+This guide demonstrates efficient GPU implementations of **triangular matrix multiplication**, as used in **causal self-attention** in autoregressive transformer models.
```
nit: "guide" -> "pocket reference"

nerdai replied:
Perhaps "provides" rather "demonstrates"?

nerdai commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
-# Kernels for triangular matmul forward pass
+<!-- markdownlint-disable-file MD033 -->
+<!-- markdownlint-disable MD013 -->
```
I think this is length rule. Can we not ignore :)

nerdai commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+## Introduction
+
+This guide demonstrates efficient GPU implementations of **triangular matrix multiplication**, as used in **causal self-attention** in autoregressive transformer models.
+For **causal (autoregressive) attention**, we only need the **lower triangle** of the attention matrix; that is, each token should only attend to current and previous tokens.
```
I think we can do "attention matrix. That is, ..." for better flow.

nerdai commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+For **causal (autoregressive) attention**, we only need the **lower triangle** of the attention matrix; that is, each token should only attend to current and previous tokens.
+
+Computing the full matrix is wasteful when only the lower triangle is needed.
+Triangular matrix multiplication is a specialized form of matrix multiplication where, instead of computing the full output matrix, only the **lower triangle** is computed. This leads to substantial computational savings.
```
"...matrix multiplication, where ..."

emersodb replied:
Perhaps worthwhile having an illustration of what we're trying to accomplish mathematically? That is, something like 
```
\begin{align*}
\begin{bmatrix}
    x_{1,1} & x_{1,2} & x_{1,3} \\
    x_{2,1} & x_{2,2} & x_{2,3} \\
    x_{3,1} & x_{3,2} & x_{3,3} \\
\end{bmatrix} \oslash \begin{bmatrix}
    y_{1,1} & y_{1,2} & y_{1,3} \\
    y_{2,1} & y_{2,2} & y_{2,3} \\
    y_{3,1} & y_{3,2} & y_{3,3} \\
\end{bmatrix}  = \begin{bmatrix}
    \displaystyle{\sum_{i=1}^3 x_{1,i}y_{i,1}} & 0 & 0 \\
    \displaystyle{\sum_{i=1}^3 x_{2,i}y_{i,1}} & \displaystyle{\sum_{i=1}^3 x_{2,i}y_{i,2}} & 0 \\
    \displaystyle{\sum_{i=1}^3 x_{3,i}y_{i,1}} & \displaystyle{\sum_{i=1}^3 x_{3,i}y_{i,3}} & \displaystyle{\sum_{i=1}^3 x_{3,i}y_{i,3}}\\
\end{bmatrix}
\end{align*}
```
![image](https://github.com/user-attachments/assets/6081f6cd-9993-4452-80bb-6123febffd10)
Just a thought. Also, I chose \oslash as the operator but if anyone knows of an extant symbol for it, feel free to not use that.

nerdai commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+Computing the full matrix is wasteful when only the lower triangle is needed.
+Triangular matrix multiplication is a specialized form of matrix multiplication where, instead of computing the full output matrix, only the **lower triangle** is computed. This leads to substantial computational savings.
+
+This guide explains a series of CUDA kernel implementations for the **Trimat forward pass**, based on the [llm.c](https://github.com/karpathy/llm.c/tree/master/dev/cuda) GitHub repository.
```
nit: "Trimat Forward Pass"

nerdai commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+
+The input tensor packs queries and keys (and values, though unused here) in the shape:
+
+(B, T, 3, NH, HS)
```
Should this be centered math equation?

$$
(B, T, 3, NH, HS),
$$

where:

nerdai replied:
\\(B\\): Batch size

nerdai commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+- NH: Number of attention heads
+- HS: Head size (HS = C / NH, where C is the total channel size)
+
+Only the Q and K portions of the input are used in this computation.
```
\\(Q\\) and similarily for K

nerdai commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+
+The output tensor has shape:
+
+(B, NH, T, T)
```
Same comment as before. Should this be a centered math equation?

nerdai commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+
+$$\text{out}[b][h][i][j] = \frac{Q[b][i][h] \cdot K[b][j][h]}{\sqrt{\text{HS}}} \quad \text{for } j \leq i$$
+
+That is, for each batch $b$, head $h$, and timestep pair $(i, j)$ such that $j <= i$, we compute the dot product between query vector $Q[b][i][h]$ and key vector $K[b][j][h]$.
```
In-line math uses "\\(<some-math>\\)" rather than "$<some-math>$"

nerdai commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+  <img src="figures/trimat_forward/kernel4.svg" alt="Kernel 4 Diagram" width="700"/>
+</center>
+
+## References
```
For consistency, can we use the same format that @emersodb uses in his pocket refs?

https://github.com/VectorInstitute/ai-pocket-reference/blob/main/books/fl/src/horizontal/vanilla_fl/fedsgd.md

nerdai reviewed (commented):
Thanks @kohankhaki for the contribution! I think there's a lot to like here. In this first pass, I've noticed there's some styling—in particular, math styling—that we need to clean up.

Also, I know its a bit annoying but the line-length rule is set to 79 and for now shouldn't be ignored. This makes .md files easier to read when viewing them as raw files. Happy to show you how I make format this to 79 line length...

emersodb commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+This guide explains a series of CUDA kernel implementations for the **Trimat forward pass**, based on the [llm.c](https://github.com/karpathy/llm.c/tree/master/dev/cuda) GitHub repository.
+These kernels avoid unnecessary computation and offer potential speedups over cuBLAS. These kernels are introduced in increasing order of optimization:
+
+- [Kernel 1: `matmul_tri_naive`](#kernel-1-naive-implementation-matmul_tri_naive): a simple nested loop implementation with no memory optimization.
```
I think it's typical to have a capital after colons in lists like this? That is "A simple..." or "Uses register..." etc.

emersodb commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+- [Kernel 3: `matmul_tri3`](#kernel-3-vectorized-loads): adds **vectorized memory access** using `float4` to improve memory coalescing.
+- [Kernel 4: `matmul_tri4`](#kernel-4-shared-memory-tiling-matmul_tri4): leverages **shared memory** tiling for inter-thread data reuse and further performance gains.
+
+The next section, [Input, Output, and Computation](#input-output-and-computation), describes the tensor shapes, the configuration used for benchmarking, and the exact computation performed during the Trimat forward pass.
```
Since you're not reporting benchmarking here (if I missed it please correct me), maybe just say these are "the configurations used in the examples,..."?

emersodb replied:
I don't think you have to capitalize trimat throughout (after introducing the name Trimat Forward Pass above, perhaps), since it's not a persons name?

emersodb commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+- T: Sequence length
+- 3: Stacked Query, Key, and Value vectors
+- NH: Number of attention heads
+- HS: Head size (HS = C / NH, where C is the total channel size)
```
If channel is the term used by, for example, in the llm.c library, let's leave this, but in my head, the more common term here is total "hidden dimension" size or perhaps "embedding" size. Again, if channels and C are common in this setting, let's stick with it.

kohankhaki replied:
The repo uses channel mostly. That's why I kept it like that.

emersodb replied:
Cool. Let's keep it 🙂

emersodb commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+
+### Configuration Used
+
+For benchmarking and validation, the implementation uses the following fixed configuration:
```
Again, maybe instead of "For benchmarking and validation," we just say. "For this example, "?

emersodb commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+
+The goal is to compute the scaled dot-product attention score between queries and keys:
+
+$$\text{out}[b][h][i][j] = \frac{Q[b][i][h] \cdot K[b][j][h]}{\sqrt{\text{HS}}} \quad \text{for } j \leq i$$
```
Nothing you can do about this, but the way md renders the sqrt here makes me sad.

emersodb replied:
You could use (\text{HS})^{\frac{1}{2}} I guess, but up to you.

emersodb commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+
+Below is a visualization of how threads compute 8×8 blocks in the output matrix:
+
+<center>
```
I think we can make this diagram a 100% width diagram rather than something fixed width so that it scales?

emersodb replied:
This is might end up being a dumb question, but the indexing on T within the blocks goes up to 15 (i.e. 16 values). Is that in conflict with the 8x8 description you have?

kohankhaki replied:
@emersodb Sorry though I had replied to this, I guess it is not saved. Thread dimension is 16x16. Each thread works on a 8x8 region.

emersodb replied:
Ah I see. That makes sense! Thanks!

emersodb commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+
+See **Figure 2** for a visualization of how registers are used to tile the data within a thread:
+
+<center>
```
I think we can make this diagram a 100% width diagram rather than something fixed width so that it scales?

emersodb commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+
+See **Figure 4** for an illustration of shared memory tiling and accumulation:
+
+<center>
```
I think we can make this diagram a 100% width diagram rather than something fixed width so that it scales?

emersodb commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+- **Each thread loads Q[i] and K[j]** directly from global memory
+- Computes **64 dot products** per thread (8 queries × 8 keys)
+- Causal masking is enforced by skipping blocks where j > i
+- **Upper triangle is ignored**, though some redundant work may occur inside diagonal blocks
```
In other, longer bullet point sections, you've ended sentences in periods. I'd say do so again here 🙂

emersodb commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+
+### Key Characteristics of Kernel 2
+
+- One thread per **8×8 tile**, same as Kernel 1
```
Similar recommendation to above. Perhaps end these bullets in periods.

emersodb commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+
+This kernel improves performance by leveraging **register tiling**. Each thread still computes an **8×8 tile** of the output, but instead of reading query and key vectors from global memory multiple times, each thread loads its Q and K vectors into registers for reuse.
+
+### Key Characteristics of Kernel 2
```
For presentation consistency, here and in Kernel 3, with Kernel 1 and Kernel 4, perhaps include the name of the kernel in parentheses like 1 and 4?

emersodb commented on books/compute/src/cuda/kernels/trimat_forward.md:
```diff
@@ -1,11 +1,153 @@
+
+### Key Characteristics of Kernel 3
+
+- Each thread still computes an 8×8 tile (64 dot products)
```
Perhaps periods again here 🙂

emersodb reviewed (commented):
Overall, I think the write up is concise and clear, especially considering the complexity of the topic. Most of my suggestions are fairly cosmetic rather than correcting any issues. 

As an aside: @kohankhaki, did you make the svg diagrams? I think they are really nice.

kohankhaki commented:
@nerdai Thank you for reviewing this. I have fixed all your comments including the length issues. 
@emersodb Thanks David for the comments. I have fixed these as well.
I created the figures using excalidraw.

kohankhaki committed:
Fixed issues.

kohankhaki committed:
enabled MD013

kohankhaki committed:
fixed trimat spelling.

emersodb reviewed (approved):
Changes look good to me. There is a single comment that is unresolved from my previous review, just asking about the indexing of the blocks in the diagrams. It's probably a naive question. If so, just ignore it and resolve 🙂. I was just curious.

nerdai commented:
@kohankhaki Excalidraw for the win!

nerdai committed:
hosted images

nerdai reviewed (approved):
@kohankhaki: Thanks for making the changes. I think the Pocket Ref reads great!

NOTE: I've swapped the local image versions for hosted ones, which is one of the final steps for preparing the Pocket Ref to be released.

kohankhaki commented:
> Changes look good to me. There is a single comment that is unresolved from my previous review, just asking about the indexing of the blocks in the diagrams. It's probably a naive question. If so, just ignore it and resolve 🙂. I was just curious.



> Changes look good to me. There is a single comment that is unresolved from my previous review, just asking about the indexing of the blocks in the diagrams. It's probably a naive question. If so, just ignore it and resolve 🙂. I was just curious.

Thank you David, just replied. :)
